server/routes/nl.py

```diff
@@ -101,21 +108,41 @@ def _remove_places(query, places_found):
 
 def _infer_place_dcid(places_found):
   if not places_found:
+    logging.info("places_found is empty. Nothing to retrieve from Maps API.")
     return ""
 
   place_dcid = ""
-  place = _maps_place(places_found[0])
-  # If maps API returned a valid place, use the place_id to
-  # get the dcid.
-  if place and ("place_id" in place):
-    place_id = place["place_id"]
-    logging.info(f"MAPS API found place with place_id: {place_id}")
-    place_ids_map = _dc_recon([place_id])
-
-    if place_id in place_ids_map:
-      place_dcid = place_ids_map[place_id]
-
-  logging.info(f"DC API found DCID: {place_dcid}")
+  for p_str in places_found:
+    # If this is a special place, return the known DCID.
+    if p_str.lower() in constants.SPECIAL_PLACES_TO_GEOIDS:
+      place_dcid = constants.SPECIAL_PLACES_TO_GEOIDS[p_str.lower()]
+      logging.info(
+          f"{p_str} was found in SPECIAL_PLACES_TO_GEOIDS. Returning its DICD {place_dcid} without querying Maps API."
+      )
+      break
+
+    # Iterate over all the places until a valid place DCID is found.
+    logging.info(f"Searching Maps API with: {p_str}")
+    place = _maps_place(p_str)
+    # If maps API returned a valid place, use the place_id to
+    # get the dcid.
+    if place and ("place_id" in place):
+      place_id = place["place_id"]
+      logging.info(f"MAPS API found place with place_id: {place_id}")
+      place_ids_map = _dc_recon([place_id])
+
+      if place_id in place_ids_map:
+        place_dcid = place_ids_map[place_id]
+        logging.info(f"DC API found DCID: {place_dcid}")
+        break
+      else:
+        logging.info(
+            f"Maps API found a place {place_id} but no DCID match found.")
```

Comment:
I think we want to somehow propagate this to debug-info eventually?  Can have TODO?

Reply:
added a TODO. Also added another TODO where we currently set the place object to None when no place dcid is found even though the places_found could be non-empty. The side-effect of that is that even if we have places_found from this NER + heuristic but did not get to a DCID, the debugInfo shows no places_found which causes confusion.